File: Mpdfin/Mpd/CommandHandler/Status.cs
using System.Diagnostics;
using LibVLCSharp.Shared;
using Microsoft.VisualBasic;

namespace Mpdfin.Mpd;

partial class CommandHandler
{
    Response Status()
    {
        Response response = new();

        response.Add("repeat"u8, "0"u8);
        response.Add("random"u8, "0"u8);
        response.Add("single"u8, "0"u8);
        response.Add("consume"u8, "0"u8);

        if (Player.CurrentPos is not null)
        {
            response.Add("song"u8, Player.CurrentPos.Value.ToU8String());
            response.Add("songid"u8, Player.CurrentSong!.Value.Id.ToU8String());
        }

        response.Add("volume"u8, Player.Volume.ToU8String());
        response.Add("state"u8, Player.State switch
        {
            VLCState.Playing => "play"u8,
            VLCState.Paused => "pause"u8,
            _ => "stop"u8,
        });
        response.Add("playlist"u8, Player.PlaylistVersion.ToU8String());
        response.Add("playlistlength"u8, Player.Queue.Count.ToU8String());
        response.Add("elapsed"u8, Player.Elapsed.ToU8String());
        response.Add("time"u8, $"{(int)Player.Elapsed}:{Player.Duration}");
        response.Add("duration"u8, Player.Duration.ToU8String());

        return response;
    }

    Response CurrentSong()
    {
        var currentSong = Player.CurrentSong;
        if (currentSong is not null)
        {
            return currentSong.Value.GetResponse(Player.CurrentPos);
        }
        else
        {
            return new();
        }
    }

    Response Stats()
    {
        Response response = new();

        response.Add("artists"u8, Db.Items.SelectMany(item => item.GetTagValue(Tag.Artist)!).Distinct().Count().ToU8String());
        response.Add("albums"u8, Db.Items.SelectMany(item => item.GetTagValue(Tag.Album)!).Distinct().Count().ToU8String());
        response.Add("songs"u8, Db.Items.Count.ToU8String());

        return response;
    }

    async Task<Response> Idle(ClientStream stream, List<string> args)
    {
        var subsystems = args.Count > 0
            ? args.Select(arg => Enum.Parse<Subsystem>(arg, true)).ToArray()
            : null;

        using CancellationTokenSource source = new();

        var notificationTask = NotificationsReceiver.WaitForEvent(subsystems, source.Token);
        var incomingCommandTask = stream.ReadRequest(source.Token);

        var finishedTask = await Task.WhenAny(notificationTask, incomingCommandTask);
        source.Cancel();

        if (finishedTask == notificationTask)
        {
            Response response = new();

            foreach (var subsystem in notificationTask.Result)
            {
                response.Add("changed"u8, Enum.GetName(subsystem));
            }

            return response;
        }
        else if (finishedTask == incomingCommandTask)
        {
            if (incomingCommandTask.Result?.Command == Command.noidle || incomingCommandTask.Result is null)
            {
                return new();
            }
            else
            {
                throw new Exception($"Only `noidle` can be called when idling, got `{incomingCommandTask.Result?.Command}`");
            }
        }
        else
        {
            throw new UnreachableException();
        }
    }
}

File: src/mpd/server/commands/status.rs
use super::CommandContext;
use crate::mpd::{error::Error, server::read_request, Request, Response, Result, Subsystem};
use futures_lite::future;
use std::borrow::Cow;
use strum::VariantArray;

pub fn status(ctx: CommandContext<'_>) -> Response {
    Response::new()
        .field("volume", ctx.server.player.volume())
        .field("repeat", 0)
        .field("random", 0)
        .field("single", 0)
        .field("consume", 0)
        .field("playlist", 0)
        .field("playlistlength", 0)
}

pub fn current_song(ctx: CommandContext<'_>) -> Result<Response> {
    Ok(Response::new())
}

pub async fn idle(ctx: CommandContext<'_>) -> Result<Response> {
    let mut buf = String::with_capacity("noidle\n".len());

    let subsystems = if ctx.args.is_empty() {
        Cow::Borrowed(Subsystem::VARIANTS)
    } else {
        let vec = ctx
            .args
            .iter()
            .map(|arg| Subsystem::try_from_str(arg))
            .collect::<Result<Vec<_>>>()?;
        Cow::Owned(vec)
    };

    enum IdleInterruption<'a> {
        Request(Result<Option<Request<'a>>>),
        Notification(Vec<Subsystem>),
    }

    let interruption = future::or(
        async { IdleInterruption::Request(read_request(ctx.stream, &mut buf).await.transpose()) },
        async { IdleInterruption::Notification(ctx.subsystem_listener.listen(&subsystems).await) },
    )
    .await;

    match interruption {
        IdleInterruption::Request(request) => match request? {
            Some(request) => {
                if request.command == "noidle" {
                    Ok(Response::new())
                } else {
                    Err(Error::InvalidArg(
                        "Only the 'noidle' command is allowed when idling".to_owned(),
                    ))
                }
            }
            None => Ok(Response::new()),
        },
        IdleInterruption::Notification(subsystems) => {
            Ok(Response::new().repeated_field("changed", &subsystems))
        }
    }
}
